File: 739-daily-temperatures/daily-temperatures.js
/**
 * @param {number[]} temperatures
 * @return {number[]}
 */
var dailyTemperatures = function(temperatures) {
    let ans = new Array(temperatures.length).fill(0);
    let stack = [];

    for(let i = temperatures.length-1;i>=0;i--){
        while(stack.length && temperatures[stack[stack.length-1]]<=temperatures[i]){
            stack.pop();
        }

        if(stack.length && temperatures[stack[stack.length-1]] > temperatures[i]){
            ans[i] = stack[stack.length -1] - i;
        }
        stack.push(i);
    }
    return ans;
};

File: 739-daily-temperatures/daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        vector<int> ans(temperatures.size(), 0);
        vector<int> stack;

        for(int i = temperatures.size()-1;i>=0;i--){
            while(!stack.empty() && temperatures[stack.back()]<=temperatures[i]){
                stack.pop_back();
            }

            if(!stack.empty() && temperatures[stack.back()] >temperatures[i]){
                ans[i] = stack.back() - i;
                stack.push_back(i);
            }
            stack.push_back(i);
        }

        return ans;
    }
};